### ISSUE TITLE
Trailing slashes - tracked as two separate pages

### ISSUE THREAD

**Issue by agendartobias** [piwik/piwik#17055]
A given page is being tracked as two separate pages due to trailing slashes in the URL. For example:

/login
/login/

Is there any way to make Matomo recognize that the two URLs are actually a single page?

**Comment by mobiliodevelopment** [comment]
Actually this are two different URLs and should be counted independent.

Example:
/login -> /login
/login/ -> /login/index.html or index.htm

**Comment by tsteur** [comment]
Hi @agendartobias AFAIK there isn't a way yet to recognise the two URLs as a single page. It's unlikely we'll work on this as it is technically two different URLs. However, I'll keep the issue open as someone should be able to implement this feature in a plugin without too much effort. Happy to guide them in the right direction if someone is keen.

**Comment by dl-lim** [comment]
I agree with this issue. It is skewing how the data is presented when we pull from this to other BI solutions, unless plenty of transformation is being performed at a granular level.

> Actually this are two different URLs and should be counted independent.
> 
> Example:
> /login -> /login
> /login/ -> /login/index.html or index.htm

If this is the case, then /login/ should already be logged as /login/index.html . What is the point of logging /login/ then?

By this, shouldn't `/` be the largest then, since ALL visits go thru `/` ? - which actually isn't the case in our data. It appears to be inconsistent where this is applied.

**Comment by artus70** [comment]
The URL /login _could_ mean the same as /login/, but it doesn't have to. Therefore, I plead for it to become **configurable**.

By being able to combine it, cases like this (same URL as single page and as directory) could be avoided:

![grafik](https://user-images.githubusercontent.com/367315/122954367-a8216400-d37f-11eb-9d26-36be56e9917a.png)

(I'd love to have the upper URL shown/positioned as the index page of the lower URL instead of being shown separately!)

**Comment by vladimir-shmidt** [comment]
Usually server treat /login and /login/ the same. index.html or index.php is the next level of discussion but there is a good approach to combine (group) such requests and treat as the same url

I wonder have matomo some ability or extension point to solve it? 

I thought it is SPA problem with router but haven't found anything criminal - client can put what ever he want's like login and login/ and login/?variable=value even application don't have any variables. So client redirection can solve trailing slash a bit but it is much easy to group queries in matomo statistics for me (in case there is a way to do it :) )

**Comment by tassoman** [comment]
>  I'll keep the issue open as someone should be able to implement this feature in a plugin without too much effort. Happy to guide them in the right direction if someone is keen.

@tsteur I'm interested on such kind of aggregation, but I can't understand if it can be done at tracking time (maybe a plugin extending the tracker?) or if can be done at reporting time (maybe a plugin working during report creation job) 🤔 

Any idea about it? Are there example plugins working at tracking time or reporting time?

**Comment by artus70** [comment]
> at tracking time

I've no working example but this approach would be okay for me, i.e. by setting this in the JS of the pages:

`_paq.push(['treatAsDirectory', 1]);`

By doing it this way it's optional, as I suggested before (https://github.com/matomo-org/matomo/issues/17055#issuecomment-866085762).

The Matomo Javascript could then save it as if it had a trailing slash. At least on my site, it's no difference if I call a page with or without trailing slash. In case a slash is given we automatically redirect the visitor to the URL without the slash.

I guess this approach would be the easiest to implement.

Of course Matomo in addition could treat the above mentionend JS setting to save this information someware internally to be able to generate the corresponding URLs (without the slash) in its output. But for me, that's not a must.
